### ISSUE TITLE
Release sVim as official extension

### ISSUE THREAD

**Issue by nikitavoloboev** [flipxfx/sVim#94]
Sadly with new release of macOS Mojawe, [non official Safari extensions will break](https://www.howtogeek.com/fyi/macos-mojave-will-break-a-bunch-of-safari-extensions/).

I am planning to update my mac to Mojawe soon but I can't do it until I know that sVim is working in new Safari as I can't use the browser without this extension.

I know it'd dumb for developers having to pay 100 dollars just to release an extension in their store but perhaps if the community of users can donate to cover the costs of @flipxfx or someone else publishing the extension, it would be worth doing it.

What do you think?

**Comment by nikitavoloboev** [comment]
Although it seems it [might still be possible to load extensions](https://georgegarside.com/blog/macos/install-any-safari-extension-macos-mojave/) by bypassing the system.

**Comment by nikitavoloboev** [comment]
Oh and to add to this, it seems sVim alongside all other extensions will have to be rewritten with Cocoa instead of plain JavaScript. Hopefully that transition will not be too painful though.

**Comment by nikitavoloboev** [comment]
Anyone wants to give a hand at porting sVim to Mojawe and new Safari?

**Comment by nikitavoloboev** [comment]
I agree with you @abrookins. But even knowing all this, Safari is still the best browser on macOS right now.

And I really need sVim to work there. I am going to try replicate this extension myself and hopefully it should be possible.

**Comment by abrookins** [comment]
I'm not sure I'll ever be able to escape the Apple ecosystem. 😅 Has anyone found an overview of the update we need to make? Am I correct that this needs to become an "App Extension"? https://developer.apple.com/documentation/safariservices/safari_app_extensions CC @nikitavoloboev

**Comment by nikitavoloboev** [comment]
Yes it's correct. 

It has to be rewritten as a whole basically.

**Comment by nikitavoloboev** [comment]
> I have already a migrated version running on default settings and without control to safari tabs.

Can you make a fork or start a repo with it? I am planning to move to Mojawe when it gets released.

**Comment by luchenyuxx** [comment]
@nikitavoloboev 
I will commit it to a new branch on my fork.

**Comment by abrookins** [comment]
If you push it up to a branch somewhere, I’ll test it with Mojave this
weekend. Then maybe we can get an early Mojave beta out. I’ll probably
renew my dev membership and we can use that to release (that is, if you
wanted to contribute that code to sVim rather than fork the project). Also,
awesome that you already jumped on the update @luchenyuxx
<https://github.com/luchenyuxx>!

**Comment by abrookins** [comment]
I’ll start looking at how to do the plugin settings.

**Comment by abrookins** [comment]
Alright, I renewed, so now we have to release this thing. 😅

**Comment by luchenyuxx** [comment]
@abrookins @nikitavoloboev 
The migrated Xcode project: [link](https://github.com/luchenyuxx/sVim/tree/app).
Use Xcode to open it. Follow the official guide to launch and use it in Safari.

**Comment by abrookins** [comment]
Just to expand on this a bit, here are some of my thoughts:

- I updated to Mojave and installed @luchenyuxx's branch. Great to see that hints and navigation work without too much rewriting.

- We can adapt the JS portion of the app to send messages wherever it previously interacted directly with the JS Safari API. However, as @luchenyuxx noted already, the Swift/Obj-C Safari API exposed to the app code has less features than the JS API, so the app version of sVim will have less features, too.

- Reading a bit closer on the Safari app extension timeline, it appears that Apple is deprecating support for side-loaded Safari extensions, but will *still support* versions of sVim released from the Safari Extensions Gallery, which sVim is currently. However, they will stop accepting gallery submissions at the end of 2018. So, for the immediate future, the plugin should continue to work in Mojave, if installed from the extensions gallery (and I've confirmed that it does in the Mojave beta).

Given all that, here's what I propose:

- Before the end of the year, we make another release or two. That should get a couple of recent PRs merged and available in the build from the extensions gallery.

- There will be a period of time, I suspect about 9 months (let's call it January 2019 - October 2019), while the extensions gallery is not accepting new submissions, and during which our last extensions gallery release should still be installable. During that time, we can work on releases of a new thing, "sVim2", in the app store.

- Apple will eventually deprecate extensions gallery plugins completely, so our frozen version from 2018 will disappear. Whenever the happens, the new app version, sVim2, will already have been released and will become the only way to install sVim in Safari.

**Comment by nikitavoloboev** [comment]
Is it possible to install @luchenyuxx fork officially somehow? I am not updating to Safari 12.0 until sVim is available there.

**Comment by abrookins** [comment]
It's kinda buried in my wall of text, but the summary is that the current extension should work fine in Mojave. What's changing is that end of year, Apple will stop accepting updates to the extensions gallery. We'll then probably have the rest of 2019 to get the app version working, and I hope by then they have better Safari APIs available.

**Comment by nikitavoloboev** [comment]
But the extension is not on Safari Extension List. It's downloaded from releases currently.

How can I install this extension on Safari 12.0?

**Comment by abrookins** [comment]
Should be right here:
https://safari-extensions.apple.com/details/?id=com.flipxfx.svim-6Q2K7JYUZ6

**Comment by nikitavoloboev** [comment]
Is there any codebase I can look that has partial support of sVim features for new Safari?

**Comment by muyinliu** [comment]
> Should be right here: https://safari-extensions.apple.com/details/?id=com.flipxfx.svim-6Q2K7JYUZ6
> […](#)

Can NOT find sVim...

**Comment by translit** [comment]
> Can NOT find sVim...

That's because that URL now redirects to the App Store and sVim hasn't been published there so far.

**Comment by muyinliu** [comment]
> > Can NOT find sVim...
> 
> That's because that URL now redirects to the App Store and sVim hasn't been published there so far.

Thanks. I spent some time to recover Safari 12 from TimeMachine for macOS 10.13.6, related files:

```
.
├── Applications
│   └── Safari.app
├── Library
│   └── Widgets
│       └── Web\ Clip.wdgt
├── System
│   └── Library
│       ├── Accounts
│       │   └── Notification
│       │       └── CloudBookmarksAccountsNotifier.bundle
│       ├── CoreServices
│       │   └── SafariSupport.bundle
│       ├── Frameworks
│       │   └── SafariServices.framework
│       ├── LaunchAgents
│       ├── Preferences
│       │   └── Logging
│       │       └── Subsystems
│       │           ├── com.apple.Safari.plist
│       │           └── com.apple.SafariShared.plist
│       ├── PrivateFrameworks
│       │   └── SafariFoundation.framework
│       └── StagedFrameworks
│           └── Safari
│               ├── JavaScriptCore.framework
│               ├── Safari.framework
│               ├── SafariShared.framework
│               ├── WebCore.framework
│               ├── WebDriver.framework
│               ├── WebInspector.framework
│               ├── WebInspectorUI.framework
│               ├── WebKit.framework
│               ├── WebKitLegacy.framework
│               └── libwebrtc.dylib
└── usr
    └── share
        └── man
            ├── man1
            │   └── safaridriver.1
            └── man8
                ├── SafariBookmarksSyncAgent.8
                ├── SafariCloudHistoryPushAgent.8
                ├── SafariHistoryServiceAgent.8
                ├── SafariLaunchAgent.8
                ├── SafariNotificationAgent.8
                ├── SafariPlugInUpdateNotifier.8
                └── webinspectord.8
```

Hope it would be helpful. All related path comes from [Safari12.1.2HighSierraAuto.pkg](https://files02.tchspt.com/storage2/temp/Safari12.1.2HighSierraAuto.pkk) with tool [Pacifist](https://www.charlessoft.com)

**Comment by flippidippi** [comment]
Hey everyone!

Sorry I've neglected sVim so much these past years. It was something I was super excited about, life happened, I got busy, and didn't like the direction that Apple was going with their Safari extensions. I really do miss using sVim and I'm ready to get back at it, bite the bullet, and pay for an Apple developer license. This is my first time building an official Safari extension "the new way" so it'll take me a while to get it going, but I'm determined to get sVim up and running again. Also, from what I've seen with new extensions it might be better to just do a full rewrite, which will also add some time. While I can't give a timeline on when it'll be ready, I just wanted everyone to know I'll start back at it again soon 😄 

@abrookins

**Comment by nikitavoloboev** [comment]
> While I can't give a timeline on when it'll be ready, I just wanted everyone to know I'll start back at it again soon 😄

Thank you 💚

Please do apply for GitHub Sponsors or create a Patreon. I am willing to donate to support this effort and pay off the license fee. I am sure I am not alone. 😊

**Comment by translit** [comment]
>  I'm ready to get back at it

Thanks so much. After sVim stopped working, I switch to Vimari, but it lacks my beloved ```gi```.

**Comment by de-Blaxe** [comment]
@flippidippi I’ve moved a lot of your extension into the new safari app extension in a branch of my fork. You’re welcome to jump off of that!

**Comment by nikitavoloboev** [comment]
@flippidippi Have you made any progress in making sVim available for Catalina? ♥️

**Comment by nikitavoloboev** [comment]
Someone mentioned there was a working fork of sVim with limited features. Anyone has a link?

**Comment by rk1** [comment]
@nikitavoloboev The only thing I've seen mentioned before in other issues is https://github.com/de-Blaxe/sVim, but the last commit is from around 10 months ago.

I would pay for sVim running on Safari 13 or can help covering the cost of the dev program, it's the only thing holding me back from upgrading to Catalina and running an old Safari version.

**Comment by ingjieye** [comment]
vimari has a Appstore version, but seems buggy to me, would be nice to have
a Appstore version of sVim

Rossen K <notifications@github.com>于2020年1月25日 周六上午12:00写道：

**Comment by nikitavoloboev** [comment]
Vimari is not close to sVim at all. Can't even configure (rebind) keys.

**Comment by de-Blaxe** [comment]
I've just been busy with stuff, but maybe I can help around with the port until the creator comes back. The bulk of my port was just moving the old html/js into a Mac app extension and then fixing a few features that I had used frequently.

**Comment by rk1** [comment]
Agreed -- rebinding, blacklisting, gist syncing and the extensive configuration make sVim way better imo.

**Comment by nikitavoloboev** [comment]
@de-Blaxe do you have telegram or somewhere I can message you?

I had some issues building your forked extension. And want to help you move the project forward with the migration.

**Comment by muyinliu** [comment]
Any update now? It seems many days passed...

**Comment by flippidippi** [comment]
Yes. I started working on it. But I got super annoyed with how development is in Xcode. Does anyone know if there's a way to update the Javascript/CSS without Cleaning/Rebuilding the app every time?

**Comment by flippidippi** [comment]
That looks promising. Hopefully that will work better because the current developer experience is god awful. I gave up :/ I will look into this once I have the new OS.

**Comment by muyinliu** [comment]
Any update??

**Comment by flippidippi** [comment]
Unfortunately I no longer work on a Mac :/

**Comment by muyinliu** [comment]
After waiting so much time, I decide to switch to Chrome. Goodbye, Safari. Sad :(

**Comment by muyinliu** [comment]
> > I'm ready to get back at it
> 
> Thanks so much. After sVim stopped working, I switch to Vimari, but it lacks my beloved `gi`.

Vimari now support `g i` as `goToFirstInput`.
